### ISSUE TITLE
Numerical gradient check fails?

### ISSUE THREAD

**Issue by bartvm** [Element-Research/rnn#28]
So this is my first time toying around with Torch modules and the like, so there's a big chance I'm overlooking something obvious. I was trying to implement an attention model, but when testing the gradients using `optim.checkgrad` they didn't match. I later realised that even for this simple model, I can't get them to match:
```lua
nn = require 'nn'
require 'rnn'
require 'optim'

hiddenSize = 2
nIndex = 2
r = nn.Recurrent(hiddenSize, nn.LookupTable(nIndex, hiddenSize),
                 nn.Linear(hiddenSize, hiddenSize))

rnn = nn.Sequential()
rnn:add(r)
rnn:add(nn.Linear(hiddenSize, nIndex))
rnn:add(nn.LogSoftMax())

criterion = nn.ClassNLLCriterion()

function f(x)
  parameters:copy(x)
  -- Do the forward prop
  rnn:zeroGradParameters()
  local err = 0
  for i = 1, sequence:size(1) - 1 do
     local output = rnn:forward(sequence[i])
     err = err + criterion:forward(output, sequence[i + 1])
     local gradOutput = criterion:backward(output, sequence[i + 1])
     rnn:backward(sequence[i], gradOutput)
  end
  r:backwardThroughTime()
  r:forget()
  return err, grads
end

parameters, grads = rnn:getParameters()

sequence = torch.Tensor{1, 2, 1, 2}:resize(4, 1)
local err = optim.checkgrad(f, parameters:clone())
print(err)
```
This gives me errors anywhere between 0.1 and 0.01, which is way too big. After some digging I got to [these lines](https://github.com/Element-Research/rnn/blob/897c461c47ab2659246c63ea803fdba78c62c2d6/Recurrent.lua#L140-L147) in `Recurrent.lua`. Removing these lines seems to fix the problem, making the gradient error falls to around `1e-7`.
```lua
         -- startModule's gradParams shouldn't be step-averaged
         -- as it is used only once. So un-step-average it
         local params, gradParams = self.startModule:parameters()
         if gradParams then
            for i,gradParam in ipairs(gradParams) do
               gradParam:mul(rho)
            end
         end
```
I fail to see where `gradParams` get averaged, so I don't really understand the logic behind these lines. They seem to just scale the gradients for the initial hidden states with the number of steps?

**Comment by nicholas-leonard** [comment]
@bartvm You are absolutely right. I though I removed that when I removed the size-averaging. Integrated your changes and unit test here : https://github.com/Element-Research/rnn/commit/1613d03166138c8827319c7a391a631217aca1b9. Many thanks for finding this. Kind of freakish that none of the unit tests caught this o_O...

**Comment by bartvm** [comment]
Great, thanks!

So up next I tried to do the same numerical gradient check using LSTMs, but failed again. In this case, I tracked it down to `LinearNoBias`. This subclass sets the bias to zero at the beginning, and resets the gradient accumulation before the update, but if you call `model:getParameters()` it will still return the bias as trainable parameters, and they are part of the forward prop. This means that the numerical gradient estimate ends up being "wrong".

I can see the motivation behind the minimal implementation of `LinearNoBias` but given that it (a) does useless calculations and (b) the `parameters` and `getParameters` calls return parameters that should not be trained, I'd prefer a proper implementation of `LinearNoBias` instead (even though it's largely a copy paste of `Linear` with some lines removed).
